web3pi_proxy/core/rpc/node/endpoint_pool/endpoint_connection_pool.py

```diff
@@ -103,40 +103,69 @@ def __init__(
         self.status = self.PoolStatus.ACTIVE.value
         self.__lock = Lock()
         self.__logger = get_logger(f"EndpointConnectionPool.{id(self)}")
+        self.connection_close_queue: SimpleQueue[EndpointConnection | None] = SimpleQueue()
+
+        self.closing_thread = Thread(
+            target=self.__run_closing_thread,
+            daemon=True,
+        )
+        self.closing_thread.start()
+
+        self.cleanup_thread = Thread(
+            target=self.__run_cleanup_thread,
+            daemon=True,
+        )
+        self.cleanup_thread.start()
 
     def __str__(self):
         return f"{self.__class__.__name__}({self.endpoint})"
 
     class PoolStatus(Enum):
         ACTIVE = "ACTIVE"
         DISABLED = "DISABLED"
+        CLOSING = "CLOSING"
         CLOSED = "CLOSED"
 
-    def __cleanup(self) -> None:
-        excessive_connections = self.connections.qsize() - self.MAX_CONNECTIONS
-        if excessive_connections > 0:
-            self.__logger.debug(
-                f"Detected {excessive_connections} excessive connections"
-            )
-            for _ in range(excessive_connections):
-                connection = self.__get_connection()
-                self.__logger.debug(f"Removed connection {connection}")
-                try:
-                    connection.close()  # TODO: Move out of lock for better performance
-                except OSError:
-                    self.__logger.error(f"Failure on closing connection {connection}")
+    def __run_closing_thread(self):
+        while True:
+            connection = self.connection_close_queue.get()
+            if not connection:  # None is the sentinel
+                break
+            try:
+                connection.close()
+            except Exception as ex:
+                self.__logger.error(f"Error while closing a connection", ex)
+
+    def __run_cleanup_thread(self) -> None:
+        while True:
+            with self.__lock:
+                if self.status == self.PoolStatus.CLOSED.value or self.status == self.PoolStatus.CLOSING.value:
+                    break
+                if self.status == self.PoolStatus.ACTIVE.value:
+                    excessive_connections = self.connections.qsize() - self.MAX_CONNECTIONS
+                    if excessive_connections > 0:
+                        for _ in range(excessive_connections):
+                            connection = self.__get_connection()
```

Comment:
don't we need to test for an Exception here? or is this taken care of by the lock?